path_oidc.go

```diff
@@ -404,119 +393,94 @@ func (b *jwtAuthBackend) authURL(ctx context.Context, req *logical.Request, d *f
 
 	// If configured for form_post, redirect directly to Vault instead of the UI,
 	// if this was initiated by the UI (which currently has no knowledge of mode).
-	///
+	//
 	// TODO: it would be better to convey this to the UI and have it send the
 	// correct URL directly.
 	if config.OIDCResponseMode == responseModeFormPost {
 		redirectURI = strings.Replace(redirectURI, "ui/vault", "v1", 1)
 	}
 
-	provider, err := b.getProvider(config)
+	// Create a new OIDC request state
+	oidcRequest, err := b.createOIDCRequest(config, role, roleName, redirectURI, clientNonce)
 	if err != nil {
-		logger.Warn("error getting provider for login operation", "error", err)
+		logger.Warn("error generating OAuth state", "error", err)
 		return resp, nil
 	}
 
-	// "openid" is a required scope for OpenID Connect flows
-	scopes := append([]string{oidc.ScopeOpenID}, role.OIDCScopes...)
-
-	// Configure an OpenID Connect aware OAuth2 client
-	oauth2Config := oauth2.Config{
-		ClientID:     config.OIDCClientID,
-		ClientSecret: config.OIDCClientSecret,
-		RedirectURL:  redirectURI,
-		Endpoint:     provider.Endpoint(),
-		Scopes:       scopes,
+	provider, err := b.getProvider(config)
+	if err != nil {
+		logger.Warn("error getting provider for login operation", "error", err)
+		return resp, nil
 	}
 
-	stateID, nonce, err := b.createState(roleName, redirectURI, clientNonce)
+	urlStr, err := provider.AuthURL(ctx, oidcRequest)
 	if err != nil {
-		logger.Warn("error generating OAuth state", "error", err)
+		logger.Warn("error generating auth URL", "error", err)
 		return resp, nil
 	}
-	if config.NamespaceInState && len(namespace) > 0 {
-		// embed namespace in state in the auth_url
-		stateID = fmt.Sprintf("%s,ns=%s", stateID, namespace)
-	}
 
-	authCodeOpts := []oauth2.AuthCodeOption{
-		oidc.Nonce(nonce),
+	// embed namespace in state in the auth_url
+	if config.NamespaceInState && len(namespace) > 0 {
+		stateWithNamespace := fmt.Sprintf("%s,ns=%s", oidcRequest.State(), namespace)
```

Comment:
Don't we have to update the state in `oidcRequest` with this value?

Reply:
I originally thought the same, but the current behavior in the plugin doesn't update the actual state ID with the `,ns=<namespace>` suffix. You can see that the state ID is first created and set on [path_oidc.go#L483](https://github.com/hashicorp/vault-plugin-auth-jwt/blob/master/path_oidc.go#L483), which occurs before the suffix is added on [path_oidc.go#L439](https://github.com/hashicorp/vault-plugin-auth-jwt/blob/master/path_oidc.go#L439). It appears that the UI parses out the namespace suffix on [auth-jwt.js#L134](https://github.com/hashicorp/vault/blob/master/ui/app/components/auth-jwt.js#L134) to provide the proper state ID to the plugin callback handler.

Reply:
Yeah, re-visited the logic and it makes sense now!